Bug/fencing shallow volume

iPraveenParihar opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

This PR includes following - 

- This commit changes to always advertise controller
PUBLICH_UNPUBLISH capability to avoid stale volumeattachments.
Previously, the Controller PUBLISH_UNPUBLISH capability was
only advertised when the 'enable-fencing' flag was set. In
this setup, the external-attacher would add a finalizer to
the VolumeAttachment. If the user later disabled fencing by
setting enable-fencing to false, the capability would no longer
be advertised, causing the external-attacher to skip finalizer
removal. This left the VolumeAttachment stuck in the deletion phase.

- **Problem**: For a snapshot-backed PVC, there is no actual subvolume in the
Ceph cluster. Attempting to store metadata would fail as the decoded
subvolume name is not found in the cluster.
**Solution**: For a snapshot-backed PVC, store the metadata in its snapshot
using the volumeHandle to uniquely identify it. The clientaddress
metadata format is now `.cephfs.csi.ceph.com/clientaddress/<volId>/<nodeId>`.

```
Note: Previously, the client address metadata was stored as `.[cephfs|rbd].csi.ceph.com/clientaddress/<nodeId>.`
With this change, the format is updated to `.[cephfs|rbd].csi.ceph.com/clientaddress/<volId>/<nodeId>.`
If `enable-fencing` flag was enabled and volumes were staged using an older version of Ceph-CSI,
fencing may not function correctly. To restore proper fencing behavior, users may need to unstage and
re-stage those volumes.
```

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md) updated wit breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

iPraveenParihar committed:
cleanup: avoid duplication for GetControllerPublishSecret

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
helm: update templates for fencing

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
util: always advertise Controller PUBLISH_UNPUBLISH cap

This commit changes to always advertise controller
PUBLICH_UNPUBLISH capability to avoid stale volumeattachments.

Previously, the Controller PUBLISH_UNPUBLISH capability was
only advertised when the 'enable-fencing' flag was set. In
this setup, the external-attacher would add a finalizer to
the VolumeAttachment. If the user later disabled fencing by
setting enable-fencing to false, the capability would no longer
be advertised, causing the external-attacher to skip finalizer
removal. This left the VolumeAttachment stuck in the deletion phase.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
cephfs: store clientaddress metadata to snapshot-backed volume

Problem: For a snapshot-backed PVC, there is no actual subvolume in the
Ceph cluster. Attempting to store metadata would fail as the decoded
subvolume name is not found in the cluster.

Solution: For a snapshot-backed PVC, store the metadata in its snapshot
using the volumeHandle to uniquely identify it. The clientaddress
metadata format is now `.cephfs.csi.ceph.com/clientaddress/<volId>/<nodeId>`.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
rbd: include volumeId in clientaddress metadata for consistency

For CephFS read-only PVCs created from snapshots, metadata is stored
in the snapshot and identified using the volumeHandle. The clientaddress
metadata format there is:
'.cephfs.csi.ceph.com/clientaddress/<volId>/<nodeId>'

This change updates the clientaddress metadata format for RBD to also
include the volumeId, aligning it with the CephFS format:
'.rbd.csi.ceph.com/clientaddress/<volId>/<nodeId>'.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
- CephFs

```
$ k get pvc
NAME                 STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   VOLUMEATTRIBUTESCLASS   AGE
cephfs-pvc           Bound    pvc-d9ab7e6e-02b2-42a0-91ed-5b9624ced0d1   1Gi        RWO            rook-cephfs    <unset>                 111s
cephfs-pvc-restore   Bound    pvc-53a90633-9a37-4523-8f12-85412f97c131   1Gi        ROX            rook-cephfs    <unset>                 100s


$ k -it exec rook-direct-mount-5f695b75d8-gh6hj -- bash
[root@c1 /]# rbd ls repl^C
[root@c1 /]# ceph fs subvolume ls myfs csi
[
    {
        "name": "csi-vol-e380bf80-90dd-437e-896b-a544502687f7"
    }
]
[root@c1 /]# ceph fs subvolume snapshot ls myfs csi-vol-e380bf80-90dd-437e-896b-a544502687f7 csi
[
    {
        "name": "csi-snap-3cd6ea08-a03c-45a3-83f4-364b0a22b785"
    }
]
[root@c1 /]# ceph fs subvolume snapshot metadata ls myfs csi-vol-e380bf80-90dd-437e-896b-a544502687f7 csi-snap-3cd6ea08-a03c-45a3-83f4-364b0a22b785 csi
{
    ".cephfs.csi.ceph.com/clientaddress/0001-0009-rook-ceph-0000000000000001-343818e8-0704-4c3b-82f1-5682711264f2/c1": "10.244.0.1",
    "csi.storage.k8s.io/volumesnapshot/name": "cephfs-pvc-snapshot",
    "csi.storage.k8s.io/volumesnapshot/namespace": "rook-ceph",
    "csi.storage.k8s.io/volumesnapshotcontent/name": "snapcontent-cd7a8148-1f2d-4af7-a89f-6850c5198fd8"
}
```

- RBD

```
[root@c1 /]# rbd ls replicapool
csi-vol-e0a910e1-ab46-44f0-ae56-648ebb50080c
[root@c1 /]# rbd image-meta ls csi-vol-e0a910e1-ab46-44f0-ae56-648ebb50080c -p replicapool
There are 4 metadata on this image:

Key                                                                                                           Value
.rbd.csi.ceph.com/clientaddress/0001-0009-rook-ceph-0000000000000002-e0a910e1-ab46-44f0-ae56-648ebb50080c/c1  10.244.0.1
csi.storage.k8s.io/pv/name                                                                                    pvc-dbede011-47ef-4828-b010-e7e1ac5d2c58
csi.storage.k8s.io/pvc/name                                                                                   rbd-pvc
csi.storage.k8s.io/pvc/namespace                                                                              rook-ceph
[root@c1 /]#
```

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar committed:
e2e: wait for PV to be deleted instead of PVC

With Controller PUBLISH_UNPUBLISH capability, external-attacher
runs detach operation, causing PV deletion to take longer.

For static volume test, validate if the PV is deleted before
deleting backend image/subvolume.

Problem: In ephemeral volume support test, deleting a pod
instantly deletes the PVC, but PV deletion takes time, leaving
backend resources and causing stale resource errors.

Solution: Wait for PV deletion instead of PVC deletion to ensure
backend resources are fully cleaned up.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
failed [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.32/runs/341/nodes/90/log/?start=0)
```
  I0814 12:34:49.935535 81046 utils.go:2009] backend images not matching kubernetes resource count,image count 1 kubernetes resource count 0
  backend image Info:
   [csi-vol-a5647b21-0fab-48d0-8300-3b6465a8e819]
   images information and status Pool: replicapool, Image: csi-vol-a5647b21-0fab-48d0-8300-3b6465a8e819, Info: {"name":"csi-vol-a5647b21-0fab-48d0-8300-3b6465a8e819","id":"138a6aabc882","size":1073741824,"objects":256,"order":22,"object_size":4194304,"snapshot_count":0,"block_name_prefix":"rbd_data.138a6aabc882","format":2,"features":["layering"],"op_features":[],"flags":[],"create_timestamp":"Thu Aug 14 12:34:39 2025","access_timestamp":"Thu Aug 14 12:34:39 2025","modify_timestamp":"Thu Aug 14 12:34:39 2025"}
  , Status: {"watchers":[]}
 ```

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
[logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.32/runs/342/nodes/90/log/?start=0)
```
  I0814 17:04:37.880768 81868 utils.go:2009] failed to validate rbd block pv: PV Create API error: object is being deleted: persistentvolumes "pv-name" already exists
  [38;5;9m[FAILED][0m in [It] - /go/src/github.com/ceph/ceph-csi/e2e/utils.go:2010 [38;5;243m@ 08/14/25 17:04:37.881[0m
```

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

nixpanic commented on internal/cephfs/controllerserver.go:
```diff
@@ -1143,12 +1143,40 @@ func (cs *ControllerServer) ControllerPublishVolume(
+		secretName, secretNamespace, err := util.GetControllerPublishSecretRef(volumeId, util.CephFsType)
+		if err != nil {
+			log.WarningLog(ctx, "controller publish secret not found: %v", err)
+
```
please leave a comment why this is not considered an error

iPraveenParihar replied:
comment added below

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1705,12 +1705,41 @@ func (cs *ControllerServer) ControllerPublishVolume(
+		if err != nil {
+			log.WarningLog(ctx, "controller publish secret not found: %v", err)
+
+			return &csi.ControllerPublishVolumeResponse{}, nil
```
same here, explain why missing secrets is not an error

nixpanic commented on internal/cephfs/controllerserver.go:
```diff
@@ -1222,32 +1226,64 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+	ctx context.Context,
+	volOptions *store.VolumeOptions,
+	cr *util.Credentials,
+	secrets map[string]string,
```
passing both credentials and secrets look unneeded?

iPraveenParihar replied:
I can do that.

iPraveenParihar replied:
done

nixpanic commented on internal/cephfs/controllerserver.go:
```diff
@@ -1222,32 +1226,64 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
 // fenceNode attempts to fence a client node from accessing the CephFS subvolume.
 //
 // Parameters:
+//   - volumeId: The ID of the volume that may be fenced.
```
is the `volumeId` not part of the `volOptions`?

iPraveenParihar replied:
No, it is not
we have `VolOptions.VolId` which is subvolume name.

nixpanic commented on internal/cephfs/controllerserver.go:
```diff
@@ -1222,32 +1226,64 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
 // Parameters:
+//   - volumeId: The ID of the volume that may be fenced.
 //   - nodeId: The ID of the node that may be fenced.
+//   - secrets: The secrets to access the Ceph cluster.
```
Why use secrets and not the credentials anymore? Passing credentials is much cleaner.

iPraveenParihar replied:
would need to refactor the `NewSnapshotOptionsFromID` which requires both secrets and credentials

```
sVolOpts, _, sid, err := store.NewSnapshotOptionsFromID(
			ctx, volOptions.BackingSnapshotID, cr, secrets, cs.ClusterName, cs.SetMetadata,
)
```
Can be done in different PR?

nixpanic replied:
> Can be done in different PR?

Yes, please follow-up with a cleanup PR for `NewSnapshotOptionsFromID()`.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.33

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1834,7 +1834,7 @@ func (cs *ControllerServer) fenceNode(
 	}
 
-	metadataKey := getClientAddressKey(nodeId)
+	metadataKey := getClientAddressKey(rv.VolID, nodeId)
```
Why would you want to do this? Please explain the reason in the commit message, it currently only explains what it does (which is rather obvious from the code).

nixpanic commented on e2e/pod.go:
```diff
@@ -738,3 +747,147 @@ func verifyReadAffinity(
+	return nil
+}
+
+func verifyClientAddressMetadataSnapshotBacked(
```
this function is CephFS specific, maybe include that in the name, or move it to a `cephfs_helpers.go` file (if that exists).

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.33

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
passed CI 
- https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.33/detail/mini-e2e-helm_k8s-1.33/85/pipeline
- https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.33/detail/mini-e2e_k8s-1.33/92/pipeline

iPraveenParihar commented:
addressed lint failures

nixpanic commented on deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml:
```diff
@@ -55,6 +55,7 @@ spec:
             - "--pidlimit=-1"
             - "--enableprofiling=false"
             - "--setmetadata=true"
+            # - "--enable-fencing=true"
```
Why comment this out? Here, and below. Just remove the line instead and verify the documentation. Or did you not intend to push this commit (no description in the commit message)?

iPraveenParihar replied:
@nixpanic, this is part of the e2e commit. This is needed so e2e can enable it by replacing the 
`# - "--enable-fencing=true"` with `- "--enable-fencing=true"`

we have been doing the same for other flags as well.
https://github.com/ceph/ceph-csi/blob/147ba56d1782d4357ba713b1a084117c4f540971/deploy/rbd/kubernetes/csi-rbdplugin.yaml#L56-L57

iPraveenParihar replied:
added description in commit.
Thanks!

nixpanic reviewed (commented):
Looks close to complete.

My main concern is the change of the metadata key/name. Volumes that were staged with a previous version of Ceph-CSI (use the old key/name), may not function the same way as volumes that were staged with the new Ceph-CSI. Can you clearly explain in the PR description what the consequence is of upgrading Ceph-CSI while volumes are in use? This should probably be part of the pre-release-notes as well, to give users a heads-up.

Rakshith-R commented on internal/cephfs/controllerserver.go:
```diff
@@ -1226,32 +1230,69 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+	secrets map[string]string,
+) (*subvolumeMetadataHandler, error) {
+	conn := volOptions.GetConnection()
+	handler := &subvolumeMetadataHandler{}
```
create and return handler within if condition or outside.

only 
```
{
..
if backsnapshot {

return 
}

return 
}
```

is sufficient ?

iPraveenParihar replied:
Done

Rakshith-R commented on internal/cephfs/controllerserver.go:
```diff
@@ -68,6 +68,16 @@ type ControllerServer struct {
 
+// subvolumeMetadataHandler holds the metadata handling functions.
+type subvolumeMetadataHandler struct {
+	LogSubvolumeName     string
```
Add GetLogSubVolumeName() instead of directly exposing this variable.

Rakshith-R commented:
>My main concern is the change of the metadata key/name. Volumes that were staged with a previous version of Ceph-CSI (use the old key/name), may not function the same way as volumes that were staged with the new Ceph-CSI.

@nixpanic We've decided not to advertise fencing feature which is behind a flag(turned off) in 3.15.0 in last week's meeting so we should be good ?

Rakshith-R commented on internal/rbd/nodeserver.go:
```diff
@@ -435,18 +435,19 @@ func (ns *NodeServer) NodeStageVolume(
 // The connection.GetAddrs() method returns an address in the format "10.244.0.1:0/2686266785".
 // We parse this to extract just the IP address portion (e.g., "10.244.0.1") and store in metadata.
-// If '--enable-fencing' flag is set to false in CSI driver configuration, this function does nothing.
+// If the '--setmetadata' flag is set to false in CSI driver configuration or if the volume is static it does nothing.
```
'--setmetadata' -> '--enable-fencing'

iPraveenParihar replied:
Done, thanks!

nixpanic commented:
> @nixpanic We've decided not to advertise fencing feature which is behind a flag(turned off) in 3.15.0 in last week's meeting so we should be good ?

That is fine, but changes from the default should be documented. I we missed some parts of the deployment files, those changes should not be 'hidden' in an e2e commit.

iPraveenParihar committed:
e2e: test to verify the client address metadata exists

This commit adds e2e tests to verify if the client address
metadata exists in cephfs subvolume and rbd image.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
deploy: add enable-fencing flag to deployment files

This commit add enable-fencing flag to cephfs & rbd nodeplugin
and provisioner deployment yamls.

This is also needed for e2e so it can deploy driver with fencing enabled.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
Update from last review - 
- removed unwanted fields from subvolumeMetadataHandler struct 
- I have now separated deployment files changes to different commit.

PTAL @nixpanic @Rakshith-R 

Thanks

nixpanic commented on e2e/cephfs.go:
```diff
@@ -102,6 +102,7 @@ func createORDeleteCephfsResources(action kubectlAction) {
 			filename:   cephFSDirPath + cephFSProvisioner,
 			namespace:  cephCSINamespace,
 			oneReplica: true,
+			enableFencing: true,
```
how come this isn't `go fmt`'ed?

strange we manage to get through the CI jobs with this :thinking:

iPraveenParihar commented:
@Mergifyio queue

iPraveenParihar commented:
/retest ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
/retest ci/centos/k8s-e2e-external-storage/1.32

iPraveenParihar commented:
/retest ci/centos/k8s-e2e-external-storage/1.33

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.32

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.33

iPraveenParihar commented:
/retest ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/retest ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/retest ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
/retest ci/centos/k8s-e2e-external-storage/1.31

iPraveenParihar commented:
/retest ci/centos/upgrade-tests-cephfs

iPraveenParihar commented:
/retest ci/centos/upgrade-tests-rbd
